Check constraints when PUT-ing and POST-ing objects

tipabu opened the pull request.

tipabu committed:
Check constraints when PUT-ing and POST-ing objects

smerritt commented:
This helps, but there's more to it. For example, directory names in object paths can have at most 255 characters in them, but I don't believe the middleware is currently enforcing this.

tipabu committed:
Only enforce ProxyFS naming constraints in bimodal accounts

Previously, we were restricting normal Swift-only accounts
unnecessarily.

tipabu committed:
Require that each path component be less than 256 bytes

Previously, we could create objects and read objects with longer path
components, but trying to access them (or even just their parent
directories) through the filesystem API would typically lock up the
client.

tipabu commented:
Something worth remembering -- we have two fundamentally different types of constraints that we're in charge of checking:

1. all the "normal" Swift constraints, because we're acting like a proxy-server, and
2. all the *new* constraints that are imposed by ProxyFS, because it's a filesystem.

This addresses just the first kind; the directory-name-length limit is an example of the second.

I'll rebase and add a couple more patches.
